Share list initialisation

Joseph-Cashmore opened the pull request:
I've kept this PR as a draft as there is more to do to get the share list functionality working.

I may have got carried away with styling this PR, so if you want me to break it up I can do.

The share icon now reveals further list details (mostly static) where the user can share the list, delete it etc.
**This on it's own isn't very logical though, as it shouldn't be the only avenue to reveal the list information - I may need to rethink the icon?

The share/copy area either opens web share when clicked on or copies the text to clipboard
**Not 100% on what the effect is when opening web share on mobile - should the button/process be more intuitive?
***The text copied is currently fixed, and needs to be the actual list.

Commit:
font size adjustments

Commit:
padding adjustments

Commit:
updates to "completed" styling

I couldn't find a scenario with the use of a line separator that I liked, the heading is better.

Commit:
made new item form text more generic

Commit:
test fixes

Joseph-Cashmore committed:
Merge branch 'main' into ui-updates

Commit:
attempting an "isEditable" state to toggle different views

Getting some weird behaviours when item should lose focus.

Commit:
remove unnecessary style code

Commit:
test fix

Commit:
lint fixes

Commit:
new edit mode method (thanks sam!)

I found that the delay needed to be a certain length for this to work. All functions fine but not sure if there are ways of minimising the time

Commit:
fixed test

This feels like a bit of a cheat. The test isn't able to find the delete button in the current version, but if the element exists even with the class display none, it can be found

Commit:
web share or copy to clipboard if statement

far from being react appropriate, but is a start

Commit:
Styling the list details/share/delete header

I got carried away...

Commit:
share icon cursor pointer

Commit:
share icon toggles visibility of list details

Commit:
removed fixed height, changed copy to share?

samstradling commented on src/components/ListHeader/ListHeader.js:
```diff
@@ -1,26 +1,94 @@
+              type="text"
+              spellCheck="false"
+              value="https:thelist.app"
+              id="listText"
```
I see what you mean about this not being very 'react', you can do the following:

1. Where you have the useState hook, add [another hook](https://reactjs.org/docs/hooks-reference.html#useref) which is a variable that will be a reference to this component: 
   ```js
   const listText = useRef(null);
   ```

1. For this input component, pass in the ref:
   ```jsx
   <input
     ref={listText}
     ...
   />
   ```

1. `listText.current` is now a reference to this component for you to do what you need, e.g. when you want to copy the text:
   ```js
   listText.current.select();
   document.execCommand('copy');
   ```

samstradling replied:
One main advantage is that you're not relying on IDs being unique if this component gets re-used

Joseph-Cashmore replied:
Ok cool thanks! I'll have a go at implementing this!

samstradling commented on src/components/ListHeader/ListHeader.js:
```diff
@@ -1,26 +1,94 @@
+            icon={faPencilAlt}
+            className={styles.pencilIcon}
+            title="Edit list name"
+            tabIndex={0}
```
In terms of the UX, what about this icon toggling `listDetailsShown` and the share icon just immediately calling `shareList`?

Joseph-Cashmore replied:
That could work, would be a very easy swap too.

I felt like the share icon is definitely not a clear button to get the list details, this should be more intuitive!

Commit:
pencil icon now reveals list details

Share icon is now for sharing only, but if web share api isn't available the copy doesn't currently work - nothing to copy!

Joseph-Cashmore committed:
Merge branch 'main' into share-list

Commit:
make share function more "react" like

Commit:
test fixing

This test seemed to fail randomly...

samstradling commented on src/components/ListHeader/ListHeader.js:
```diff
@@ -1,26 +1,92 @@
+    } else {
+      listText.current.select();
+      document.execCommand('copy');
+    }
```
I think document.execCommand is a bit of an outdated way of doing things, I think we want:
```js
const url = 'https://thelist.app';
if (navigator.share) {
  navigator.share({
    title: 'Share Your List!',
    url,
  });
} else if (navigator.clipboard) {
  navigator.clipboard.writeText(url);
}
```

Joseph-Cashmore replied:
Gotta love some outdated js! 
Thanks I'll get this changed :)

samstradling commented on src/components/ListHeader/styles.module.css:
```diff
@@ -1,32 +1,132 @@
+
+.hideListDetails {
+  height: 0;
+  opacity: 0;
```
If this is going to work you'll also want to add an `overflow: hidden` here, otherwise the items are still there and are still clickable, they just aren't visible, which is why we get this weird behaviour:
![trimmed mp4](https://user-images.githubusercontent.com/9036408/101071534-a71c1700-3594-11eb-97d6-c38bd72da5d8.gif)

Commit:
modernising js copy to clipboard

Joseph-Cashmore committed:
Merge branch 'main' into share-list

Joseph-Cashmore committed:
Revert to main version

Test failed for some reason

samstradling commented on src/components/ListHeader/styles.module.css:
```diff
@@ -1,32 +1,133 @@
+
+/* .displayListDetails {
+  height: 100%;
+} */
```
I guess this can be removed?

samstradling commented on src/components/ListHeader/ListHeader.js:
```diff
@@ -1,26 +1,92 @@
+    if (navigator.share) {
+      navigator.share({
+        title: 'Share Your List!',
+        url: 'https:thelist.app',
```
You can use the `url` variable here

samstradling commented on src/components/ListHeader/ListHeader.js:
```diff
@@ -1,26 +1,92 @@
+          />
+          Delete List
+        </div>
+      </div>
```
I think we should consider moving all of this list detail stuff to its own component
